tests/Spec.hs

```diff
@@ -0,0 +1,55 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+module Main (main) where
+
+import Data.AdditiveGroup
+import Data.Monoid.Average
+import Data.VectorSpace
+import Test.Hspec
+import Test.Hspec.Checkers
+import Test.QuickCheck
+import Test.QuickCheck.Checkers
+import Test.QuickCheck.Classes (monoid)
+
+
+instance Arbitrary (Average Rational) where
+  arbitrary = Average <$> (getPositive <$> arbitrary) <*> arbitrary
+
+instance Arbitrary (Average (Rational -> Rational)) where
+  arbitrary = Average <$> (getPositive <$> arbitrary) <*> arbitrary
+
+instance EqProp (Average Rational) where
+  Average wx xΣ =-= Average wy yΣ = (wx==wy, xΣ==yΣ) =-= (True, True)
+
+main :: IO ()
+main =
+  hspec $
+    describe "Average" $ do
+      testBatch $ monoid (undefined :: Average Rational)
+      describe "laws for: Num" $ do
+        describe "addition" $ do
+          it "associativity" . property $ isAssoc @(Average Rational) (+)
+          it "commutative" . property $ isCommut @(Average Rational) (+)
+          it "left identity" . property $ leftId @(Average Rational) (+) 0
+          it "right identity" . property $ rightId @(Average Rational) (+) 0
+        describe "multiplication" $ do
+          it "associativity" . property $ isAssoc @(Average Rational) (*)
+          it "commutative" . property $ isCommut @(Average Rational) (*)
+          it "left identity" . property $ leftId @(Average Rational) (*) 1
+          it "right identity" . property $ rightId @(Average Rational) (*) 1
+      describe "laws for: vector space" $ do
+        it "associativity" . property $ isAssoc @(Average Rational) (^+^)
+        it "commutative" . property $ isCommut @(Average Rational) (^+^)
+        it "left identity" . property $ leftId @(Average Rational) (^+^) zeroV
+        it "right identity" . property $ rightId @(Average Rational) (^+^) zeroV
```

Comment:
You'll also need inverse laws:
```
        it "left inverse" . property $ \(a :: Average Int) -> (negateV a) ^+^ a =-= zeroV
        it "right inverse" . property $ \(a :: Average Int) -> a ^+^ (negateV a) =-= zeroV
```

Reply:
Thanks, done that. Of course these tests fail at the moment.

The only way to fix them seems to be something like
```
      negateV (Average w x) = Average (recip w) (negateV x)
```
(Note that renormalisation is **not** necessary here – dividing the weight before negating and multiplying the new weight after cancels!)

The problem is that this requires a fractional weight type. And if we want to generalise to `(*^)` it would get really troublesome – first attempt
```
      μ *^ Average w x = Average (w**μ) (x^*μ)  ??
```
But here, the normalisation doesn't cancel. And scaling by even a harmless factor like 100 would utterly blow the weight through the roof.

Any thoughts?